lib/source/defender/aws_iot_defender_api.c

```diff
@@ -367,120 +414,134 @@ static void _metricsPublishRoutine( IotTaskPool_t pTaskPool,
-        mqttError = AwsIotDefenderInternal_MqttSubscribe( acceptCallbackInfo,
-                                                          rejectCallbackInfo );
+        /* Invoke user's callback if there is. */
+        _handleApplicationCallback( AWS_IOT_DEFENDER_FAILURE_MQTT, NULL );
+        _unsubscribeMqtt( false );
+        IotSemaphore_Post( &_doneSem );
+    }
 
-        if( mqttError == IOT_MQTT_SUCCESS )
-        {
-            /* Step 3: create serialized metrics report. */
-            reportCreated = AwsIotDefenderInternal_CreateReport();
+    IotLogDebug( "Subscribe job ends." );
+}
 
-            /* If Report is created successfully. */
-            if( reportCreated )
-            {
-                /* Step 4: publish report to defender topic. */
-                mqttError = AwsIotDefenderInternal_MqttPublish( AwsIotDefenderInternal_GetReportBuffer(),
-                                                                AwsIotDefenderInternal_GetReportBufferSize() );
-
-                if( mqttError == IOT_MQTT_SUCCESS )
-                {
-                    IotLogDebug( "Metrics report has been published successfully." );
-                }
-            }
-        }
-    }
+/*-----------------------------------------------------------*/
 
-    /* If no MQTT error and report has been created, it indicates everything is good. */
-    if( ( mqttError == IOT_MQTT_SUCCESS ) && reportCreated )
-    {
-        IotTaskPoolError_t taskPoolError = IotTaskPool_CreateJob( _disconnectRoutine, NULL, &_disconnectJobStorage, &_disconnectJob );
+static void _metricsPublishRoutine( IotTaskPool_t pTaskPool,
+                                    IotTaskPoolJob_t pJob,
+                                    void * pUserContext )
+{
+    /* Unsed parameter; silence the compiler. */
+    ( void ) pTaskPool;
+    ( void ) pJob;
+    ( void ) pUserContext;
 
-        /* Silence warnigns when asserts are disabled. */
-        ( void ) taskPoolError;
-        AwsIotDefender_Assert( taskPoolError == IOT_TASKPOOL_SUCCESS );
+    IotMqttError_t mqttError = IOT_MQTT_SUCCESS;
+    bool reportCreated = false;
+    IotTaskPoolError_t taskPoolError = IOT_TASKPOOL_SUCCESS;
 
-        IotTaskPool_ScheduleDeferred( IOT_SYSTEM_TASKPOOL,
-                                      _disconnectJob,
-                                      _defenderToMilliseconds( AWS_IOT_DEFENDER_WAIT_SERVER_MAX_SECONDS ) );
-    }
-    else
+    /* Create serialized metrics report. */
+    reportCreated = AwsIotDefenderInternal_CreateReport();
+
+    /* If Report is created successfully. */
+    if( reportCreated )
     {
-        AwsIotDefenderEventType_t eventType;
+        /* Publish report to defender topic. */
+        mqttError = AwsIotDefenderInternal_MqttPublish( AwsIotDefenderInternal_GetReportBuffer(),
+                                                        AwsIotDefenderInternal_GetReportBufferSize(),
+                                                        &_mqttCallbackInfo );
```

Comment:
where was ```_mqttCallbackInfo ``` set? Should it be cleared off in ```AwsIotDefender_Stop```?

Reply:
_mqttCallbackInfo is global and initialized on line 98. It should not be cleared off as it is internal to library and contains callback processing for MQTT library async calls.